python/tank/platform/interactive_authentication.py

```diff
@@ -14,172 +14,272 @@
+        login = self._get_keyboard_input("Login", login)
+        password = self._get_password()
+        return login, password
+
+
+class UiAuthenticationHandler(AuthenticationHandlerBase):
     """
-    Common login logic, regardless of how we are actually logging in. It will first try to reuse
-    any existing session and if that fails then it will ask for credentials and upon success
-    the credentials will be cached.
-    :param login_functor: Functor that gets invoked to retrieve the credentials of the user.
+    Handles ui based authentication.
     """
-    if authentication.is_authenticated():
-        return
-
-    config_data = shotgun.get_associated_sg_config_data()
-    # We might not have login information, in that case use an empty dictionary.
-    login_info = authentication.get_login_info(config_data["host"]) or {}
 
-    # Ask for the credentials
-    session_token, login = login_functor(
-        config_data["host"],
-        login_info.get("login", get_login_name()),
-        config_data.get("http_proxy")
-    )
-
-    logger.debug("Login successful!")
-
-    # Cache the credentials so subsequent session based logins can reuse the session id.
-    authentication.cache_session_data(config_data["host"], login, session_token)
-
-
-def console_renew_session():
-    """
-    Prompts the user to enter his password on the command line to retrieve a new session token.
-    """
-    _do_authentication(_do_console_based_session_renewal)
+    def __init__(self, is_session_renewal):
+        """
+        Creates the UiAuthenticationHandler object.
+        :param is_session_renewal: Indicate if we are renewing a session.
+        """
+        self._is_session_renewal = is_session_renewal
+
+    def _do_authentication(self, hostname, login, http_proxy):
+        """
+        Pops a dialog that asks for the hostname, login and password of the user. If there is a current
+        engine, it will run the code in the main thread.
+        :param hostname: Host to display in the dialog.
+        :param login: login to display in the dialog.
+        :param http_proxy: Proxy server to use when validating credentials. Can be None.
+        :returns: A tuple of (hostname, login, session_token)
+        """
+        from .qt import login_dialog
+
+        from tank.platform import engine
+
+        if self._is_session_renewal:
+            logger.debug("Requesting password in a dialog.")
+        else:
+            logger.debug("Requesting username and password in a dialog.")
+
+        def _process_ui():
+            dlg = login_dialog.LoginDialog(
+                "Shotgun Login",
+                hostname=hostname,
+                login=login,
+                session_renewal=self._is_session_renewal
+            )
+            return dlg.result()
+
+        # If there is a current engine, execute from the main thread.
+        if engine.current_engine():
+            result = engine.current_engine().execute_in_main_thread(_process_ui)
```

Comment:
let's discuss this live. I think my original approach with the producer/consumer pattern actually avoided the issue you are pointing out. We might have to switch back to it actually.